codec-http2/src/main/java/io/netty/handler/codec/http2/InboundHttp2ToHttpAdapter.java

```diff
@@ -142,27 +167,25 @@ protected FullHttpMessage newMessage(int streamId, Http2Headers headers, boolean
      * </ul>
      * @param appendToTrailer
      * <ul>
-     * <li>{@code true} if a message {@code streamId} already exists then the headers
+     * <li>{@code true} if a message {@code stream} already exists then the headers
      * should be added to the trailing headers.</li>
      * <li>{@code false} then appends will be done to the initial headers.</li>
      * </ul>
-     * @return The object used to track the stream corresponding to {@code streamId}. {@code null} if
+     * @return The object used to track the stream corresponding to {@code stream}. {@code null} if
      *         {@code allowAppend} is {@code false} and the stream already exists.
      * @throws Http2Exception If the stream id is not in the correct state to process the headers request
      */
-    protected FullHttpMessage processHeadersBegin(ChannelHandlerContext ctx, int streamId, Http2Headers headers,
+    protected FullHttpMessage processHeadersBegin(ChannelHandlerContext ctx, Http2Stream stream, Http2Headers headers,
                 boolean endOfStream, boolean allowAppend, boolean appendToTrailer) throws Http2Exception {
-        FullHttpMessage msg = messageMap.get(streamId);
+        FullHttpMessage msg = getMessage(stream);
+        boolean release = true;
         if (msg == null) {
-            msg = newMessage(streamId, headers, validateHttpHeaders);
+            msg = newMessage(stream, headers, validateHttpHeaders);
         } else if (allowAppend) {
-            try {
-                HttpConversionUtil.addHttp2ToHttpHeaders(streamId, headers, msg, appendToTrailer);
-            } catch (Http2Exception e) {
-                removeMessage(streamId);
-                throw e;
-            }
+            release = false;
+            HttpConversionUtil.addHttp2ToHttpHeaders(stream.id(), headers, msg, appendToTrailer);
```

Comment:
@Scottmitch actually I think @blucas is right and the release = false should come after the method.